testbench/proto2rest.py

```diff
@@ -0,0 +1,180 @@
+        condition,
+        {
+            "ageDays": lambda x: ("age", x),
+            "createdBefore": lambda x: (
+                "createdBefore",
+                __postprocess_rest_date(x),
+            ),
+            "customTimeBefore": lambda x: (
+                "customTimeBefore",
+                __postprocess_rest_date(x),
+            ),
+            "noncurrentTimeBefore": lambda x: (
+                "noncurrentTimeBefore",
+                __postprocess_rest_date(x),
+            ),
+        },
+    )
+
+
+def __postprocess_rest_rule(rule):
+    return testbench.common.rest_adjust(
+        rule,
+        {
+            "condition": lambda x: (
+                "condition",
+                __postprocess_rest_condition(x),
+            )
+        },
+    )
+
+
+def __postprocess_rest_lifecycle(lc):
+    rules = lc.pop("rule", None)
+    if rules is not None:
+        lc["rule"] = [__postprocess_rest_rule(r) for r in rules]
+    return lc
+
+
+def __postprocess_rest_bucket_acl(bucket_id, acl):
+    copy = acl.copy()
+    copy["kind"] = "storage#bucketAccessControl"
+    copy["bucket"] = bucket_id
+    copy["etag"] = hashlib.md5(
+        "#".join([copy["bucket"], copy["entity"], copy["role"]]).encode("utf-8")
+    ).hexdigest()
+    return copy
+
+
+def __postprocess_rest_default_object_acl(bucket_id, acl):
+    copy = acl.copy()
+    copy["kind"] = "storage#objectAccessControl"
+    copy["bucket"] = bucket_id
+    copy["etag"] = hashlib.md5(
+        "#".join([copy["bucket"], copy["entity"], copy["role"]]).encode("utf-8")
+    ).hexdigest()
+    return copy
+
+
+def __postprocess_bucket_rest(data):
+    bucket_id = testbench.common.bucket_name_from_proto(data["name"])
+    data = testbench.common.rest_adjust(
+        data,
+        {
+            "name": lambda x: ("name", bucket_id),
+            "bucketId": lambda x: ("id", bucket_id),
+            "project": lambda x: ("projectNumber", x.replace("projects/", "")),
+            "createTime": lambda x: ("timeCreated", x),
+            "updateTime": lambda x: ("updated", x),
+            "iamConfig": lambda x: (
+                "iamConfiguration",
+                __postprocess_rest_iam_configuration(x),
+            ),
+            "encryption": lambda x: (
+                "encryption",
+                __postprocess_rest_encryption(x),
+            ),
+            "lifecycle": lambda x: (
+                "lifecycle",
+                __postprocess_rest_lifecycle(x),
+            ),
+            "acl": lambda x: (
+                "acl",
+                [__postprocess_rest_bucket_acl(bucket_id, a) for a in x],
+            ),
+            "defaultObjectAcl": lambda x: (
+                "defaultObjectAcl",
+                [__postprocess_rest_default_object_acl(bucket_id, a) for a in x],
+            ),
+        },
+    )
+    data["kind"] = "storage#bucket"
+    data["etag"] = hashlib.md5(data["updated"].encode("utf-8")).hexdigest()
```

Comment:
this changed 🤷‍♂️
```py
data["etag"] = base64.b64encode(data["updated"].encode("utf-8")).decode("utf-8")
```

Reply:
That was sneaky, good catch.  It is harmless and intentional: we need some kind of hash, does not matter which.   `hashlib.md5()` is at least fixed length, and was already used in some (maybe more) places.